metadata_layer.py

```diff
@@ -352,93 +352,492 @@ def __repr__(self):
+                        delta=1,
+                        end=last_checked,
+                    )
+                )
+
+            license_pool.last_checked = last_checked
+
+            if self.has_open_access_link:
+                license_pool.open_access = True
+            if self.default_rights_uri:
+                license_pool.set_rights_status(self.default_rights_uri)
+
+        return license_pool, is_new
+
+
+    @property
+    def has_open_access_link(self):
+        """Does this Circulation object have an associated open-access link?"""
+        return any(
+            [x for x in self.links 
+             if x.rel == Hyperlink.OPEN_ACCESS_DOWNLOAD and x.href]
+        )
+
+
+    def set_default_rights_uri(self, data_source_name, default_rights_uri=None):
+        """ TODO: should there be a way to delete default_rights_uri?
+        """
+        if default_rights_uri == None and data_source_name:
+            # We didn't get rights passed in, so use the default rights for the data source if any.
+            default = RightsStatus.DATA_SOURCE_DEFAULT_RIGHTS_STATUS.get(data_source_name, None)
+            if default:
+                self.default_rights_uri = default
+
+        for format in self.formats:
+            if format.link:
+                link = format.link
+                if self.default_rights_uri in (None, RightsStatus.UNKNOWN) and link.rel == Hyperlink.OPEN_ACCESS_DOWNLOAD:
+                    # We haven't determined rights in the constructor or the data source, but there's an
+                    # open access download link, so we'll consider it generic open access.
+                    self.default_rights_uri = RightsStatus.GENERIC_OPEN_ACCESS
+
+        if self.default_rights_uri == None:
+            # We still haven't determined rights, so it's unknown.
+            self.default_rights_uri = RightsStatus.UNKNOWN
+
+
+    def apply(self, pool, replace=None):
+        """  TODO: apply name provisional, doc string follows.
+        """
+        made_changes = False
+
+        if pool is None:
+            raise ValueError("CirculationData.apply needs a pool.")
+
+        if replace is None:
+            replace = ReplacementPolicy()
+
+        _db = Session.object_session(pool)
+        data_source = self.data_source(_db)
+
+        identifier = pool.identifier
+
+        # TODO:  had following comment in metadata.apply.  need to figure out if still relevant.
+        # now that pool uses a composite edition, we must create that edition before 
+        # calculating any links
+        # TODO:  if we do call pool.set_presentation_edition from here, watch out for circular logic.
+
+        self.set_default_rights_uri(data_source)
+        # TODO: be able to handle the case where the URL to a link changes or a link disappears.
+        link_objects = {}
+
+        for link in self.links:
+            if link.rel in [Hyperlink.OPEN_ACCESS_DOWNLOAD, Hyperlink.DRM_ENCRYPTED_DOWNLOAD]:
```

Comment:
Theoretically, no, but python doesn't completely prevent accidental setting of class properties.  Better be triple-safe.